source/common/router/router.cc

```diff
@@ -1329,6 +1329,25 @@ bool Filter::maybeRetryReset(Http::StreamResetReason reset_reason,
                           ? RetryState::Http3Used::Yes
                           : RetryState::Http3Used::No;
   }
+
+  auto upstream_request_started = false;
+  // TODO: Is this assertion worth the risk?
+  ASSERT(upstream_requests_.size() <= 2);
+  // If any request in this router has sent data to the upstream, we consider the request started.
+  for (auto& request : upstream_requests_) {
+    if (request->streamInfo()
+            .upstreamInfo()
+            ->upstreamTiming()
+            .first_upstream_tx_byte_sent_.has_value()) {
+      upstream_request_started = true;
+      break;
+    }
+  }
+  // If the retry policy is RESET_BEFORE_REQUEST and we already have
+  if (route_entry_->retryPolicy().retryOn() & RetryPolicy::RETRY_ON_RESET_BEFORE_REQUEST &&
+      upstream_request_started) {
+    return false;
+  }
   const RetryStatus retry_status = retry_state_->shouldRetryReset(
```

Comment:
I suspect we're going to want to pass upstream_request_started into should_retry_reset 

We want to handle the case if there's two policies and another one applies